src/pynxtools/nexus/nexus.py

```diff
@@ -27,16 +29,93 @@
     try_find_units,
     walk_elist,
     write_doc_string,
+    decode_or_not,
 )
 
 
+from typing import Union, Optional, Any
+import numpy as np
+
+from typing import Any, Optional
+import numpy as np
+
+from typing import Any, Optional, Union
+import numpy as np
+
+
+def decode_if_string(
+    elem: Any, encoding: str = "utf-8", decode: bool = True
+) -> Optional[Any]:
+    """
+    Decodes a numpy ndarray or list of byte objects or byte strings to strings.
+    If `decode` is False, returns the input value unchanged. Non-byte/str types
+    are returned without modification.
+
+    Args:
+        elem: A numpy ndarray, list, bytes, or any other object.
+        encoding: The encoding scheme to use. Default is "utf-8".
+        decode: A boolean flag indicating whether to perform decoding.
+
+    Returns:
+        A decoded string, a numpy array of decoded strings, a list of decoded strings,
+        or the input value if not decodable.
+        Returns None if the input is empty or invalid.
+
+    Raises:
+        ValueError: If decoding fails on bytes or numpy array elements.
+    """
+    # Early return if decoding is disabled
+    if not decode:
+        return elem
+
+    # Handle numpy arrays of bytes or strings
+    if isinstance(elem, np.ndarray):
+        if elem.size == 0:
+            return elem  # Return the empty array unchanged
+
+        if elem.dtype.kind == "S":  # Check if it's a bytes array (fixed-length strings)
+            decoded_array = np.vectorize(
+                lambda x: x.decode(encoding).rstrip("\x00")
+                if isinstance(x, bytes)
+                else x
+            )(elem)
+            return decoded_array.astype(str)  # Ensure the dtype is str
+
+        # Handle mixed-type arrays
+        if elem.dtype == object:
+            decoded_array = np.vectorize(
+                lambda x: x.decode(encoding) if isinstance(x, bytes) else x
+            )(elem)
+            return decoded_array.astype(str)  # Ensure all elements are strings
+
+    return decode_or_not(elem, encoding, decode)
+
+    # # Handle lists of bytes or strings
+    # elif isinstance(elem, list):
+    #     if not elem:
+    #         return elem  # Return an empty list unchanged
+
+    #     decoded_list = [decode_if_string(x, encoding, decode) for x in elem]
+    #     return decoded_list
+
+    # # Handle byte strings
+    # elif isinstance(elem, bytes):
+    #     try:
+    #         return elem.decode(encoding)
+    #     except UnicodeDecodeError as e:
+    #         raise ValueError(f"Error decoding bytes: {e}")
+
+    # # For other types, just return the input value unchanged
+    # return elem
+
+
 def get_nxdl_entry(hdf_info):
     """Get the nxdl application definition for an HDF5 node"""
     entry = hdf_info
     while (
         isinstance(entry["hdf_node"], h5py.Dataset)
         or "NX_class" not in entry["hdf_node"].attrs.keys()
```

Comment:
should not the attribute names be also checked if they were written by a different encoding?

Reply:
I did not encounter this so far. I have only seen string names from h5 files. Here, it also says so: https://stackoverflow.com/questions/69208885/accessing-hdf5-key-value-store-using-integer-keys

Reply:
OK

Reply:
Agreed, I think if you put a bytes object as a dataset key, h5py would complain.